File: apps/musclemap/tests/test_analytics.py
import re
import unittest
from pathlib import Path


INDEX_HTML = Path(__file__).parents[1] / "web" / "index.html"
GA4_MEASUREMENT_ID = "G-4Z9774J59Y"


class AnalyticsMarkupTests(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.html = INDEX_HTML.read_text(encoding="utf-8")

    def test_uses_neurodesk_ga4_measurement_id(self):
        self.assertIn(
            f"https://www.googletagmanager.com/gtag/js?id={GA4_MEASUREMENT_ID}",
            self.html,
        )
        self.assertIn(f"gtag('config', '{GA4_MEASUREMENT_ID}')", self.html)

    def test_ga4_config_honors_do_not_track(self):
        guarded_config = re.compile(
            r"var dnt = navigator\.doNotTrack \|\| window\.doNotTrack \|\| "
            r"navigator\.msDoNotTrack;\s*"
            r"if \(dnt !== '1' && dnt !== 'yes'\) \{\s*"
            rf"gtag\('config', '{GA4_MEASUREMENT_ID}'\);\s*\}}"
        )
        self.assertRegex(self.html, guarded_config)

    def test_cloudflare_analytics_is_absent(self):
        self.assertNotIn("static.cloudflareinsights.com", self.html)
        self.assertNotIn("data-cf-beacon", self.html)
        self.assertNotIn("Cloudflare Web Analytics", self.html)

    def test_privacy_disclosure_is_current(self):
        self.assertIn("uses Google Analytics to collect site usage metrics", self.html)
        self.assertIn(
            "Analytics is disabled when your browser sends a Do Not Track setting",
            self.html,
        )
        self.assertIn(
            "Patient images, voxel values, screenshots, and generated segmentations "
            "are not sent to Google Analytics",
            self.html,
        )


if __name__ == "__main__":
    unittest.main()

File: apps/vesselboost/scripts/test_analytics.mjs
import assert from 'node:assert/strict';
import { readFile } from 'node:fs/promises';

const html = await readFile(new URL('../web/index.html', import.meta.url), 'utf8');

assert.match(
  html,
  /googletagmanager\.com\/gtag\/js\?id=G-4Z9774J59Y/,
  'loads the Neurodesk GA4 measurement ID',
);
assert.match(
  html,
  /gtag\(['"]config['"],\s*['"]G-4Z9774J59Y['"]\)/,
  'configures the correct GA4 measurement ID',
);
assert.match(
  html,
  /navigator\.doNotTrack\s*\|\|\s*window\.doNotTrack\s*\|\|\s*navigator\.msDoNotTrack/,
  'checks browser Do Not Track settings',
);
assert.match(
  html,
  /doNotTrack\s*=\s*dnt\s*==\s*["']1["']\s*\|\|\s*dnt\s*==\s*["']yes["']/,
  'recognizes the same Do Not Track values as neurodesk.org',
);
assert.match(
  html,
  /if\s*\(!doNotTrack\)\s*\{[\s\S]*?gtag\(['"]config['"],\s*['"]G-4Z9774J59Y['"]\);?[\s\S]*?\}/,
  'guards GA4 configuration with the Do Not Track check',
);

assert.doesNotMatch(html, /cloudflareinsights|data-cf-beacon|Cloudflare Web Analytics/i);
assert.match(html, /Google Analytics collects page usage metrics/);
assert.match(html, /disabled when your browser sends a Do Not Track value of "1" or "yes"/);

console.log('Analytics and privacy checks passed.');
